core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
+
+  // technically this should be A | Sentinel.type
+  // the queue will consist of exclusively As until we shut down, when there will be one Sentinel
+  private[this] def impl[F[_]: Concurrent, A](q: Queue[F, AnyRef]): F[Channel[F, A]] =
+    (Concurrent[F].ref(0), Concurrent[F].ref(false), Concurrent[F].deferred[Unit]).mapN {
+      (leasesR, closedR, closedLatch) =>
+        new Channel[F, A] {
+
+          private[this] val LeftClosedF = LeftClosed.pure[F]
+          private[this] val FalseF = false.pure[F]
+
+          // might be interesting to try to optimize this more, but it needs support from CE
+          val sendAll: Pipe[F, A, Nothing] =
+            _.evalMapChunk(send(_))
+              .takeWhile(_.isRight)
+              .onComplete(Stream.exec(close.void))
+              .drain
+
+          // setting the flag means we won't accept any more sends
+          val close: F[Either[Channel.Closed, Unit]] =
+            closedR.getAndSet(true).flatMap { b =>
+              if (b) {
+                LeftClosedF
+              } else {
+                leasesR.get.flatMap { leases =>
+                  if (leases <= 0)
+                    q.offer(Sentinel).start.as(RightUnit)
                   else
-                    (
-                      State(values, size, None, (a, producer) :: producers, false),
-                      notifyStream(waiting).as(rightUnit) <* waitOnBound(producer, poll)
-                    )
-              }.flatten
+                    RightUnit.pure[F]
+                }
+              }
             }
-          }
 
-        def trySend(a: A) =
-          state.modify {
-            case s @ State(_, _, _, _, closed @ true) =>
-              (s, Channel.closed[Boolean].pure[F])
-
-            case s @ State(values, size, waiting, producers, closed @ false) =>
-              if (size < capacity)
-                (
-                  State(a :: values, size + 1, None, producers, false),
-                  notifyStream(waiting).as(rightTrue)
-                )
-              else
-                (s, rightFalse.pure[F])
-          }.flatten
-
-        def close =
-          state
-            .modify {
-              case s @ State(_, _, _, _, closed @ true) =>
-                (s, Channel.closed[Unit].pure[F])
-
-              case State(values, size, waiting, producers, closed @ false) =>
-                (
-                  State(values, size, None, producers, true),
-                  notifyStream(waiting).as(rightUnit) <* signalClosure
-                )
+          val isClosed: F[Boolean] = closedR.get
+
+          // there are four states to worry about: open, closing, draining, quiesced
+          // in the second state, we have outstanding blocked sends
+          // in the third state we have data in the queue but no sends
+          // in the fourth state we are completely drained and can shut down the stream
+          private[this] val isQuiesced: F[Boolean] =
```

Comment:
Should we make this `uncancelable`?

Reply:
Why would it need to be? Sometimes this is *used* in `uncancelble` contexts, but I don't think this itself needs to be `uncancelable` since it's just reading state.